core/src/env/engine/mod.rs

```diff
@@ -17,11 +17,21 @@ use crate::env::backend::{
     TypedEnv,
 };
 use cfg_if::cfg_if;
+use core::cell::RefCell;
+
+#[cfg(feature = "std")]
+use std::thread::LocalKey;
 
 pub trait OnInstance: Env + TypedEnv {
+    const INSTANCE: LocalKey<RefCell<EnvInstance>>;
+
     fn on_instance<F, R>(f: F) -> R
     where
         F: FnOnce(&mut Self) -> R;
+
+    fn on_uninitialized_instance<F, R>(f: F) -> R
+    where
+        F: FnOnce(&mut Self) -> R;
```

Comment:
Please try to find a way without modifying this core trait.
Also I don't know what the difference between `on_instance` and `on_uninitialized_instance` should be because `on_instance` has nothing to do with its initialization.